chain/src/txhashset/segmenter.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 The Grin Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Generation of the various necessary segments requested during PIBD.
+
+use std::{sync::Arc, time::Instant};
+
+use crate::core::core::hash::Hash;
+use crate::core::core::pmmr::ReadablePMMR;
+use crate::core::core::{BlockHeader, OutputIdentifier, Segment, SegmentIdentifier, TxKernel};
+use crate::error::{Error, ErrorKind};
+use crate::txhashset::{BitmapAccumulator, BitmapChunk, TxHashSet};
+use crate::util::secp::pedersen::RangeProof;
+use crate::util::RwLock;
+
+/// Segmenter for generating PIBD segments.
+#[derive(Clone)]
+pub struct Segmenter {
+	txhashset: Arc<RwLock<TxHashSet>>,
+	bitmap_snapshot: BitmapAccumulator,
+	header: BlockHeader,
+}
```

Comment:
Because `Chain::segmenter()` clones the `Segmenter` I think we should put the `BitmapAccumulator` in an `Arc`. Cloning the full accumulator each time would be quite expensive

Reply:
Oh I see what you mean here. I thought you were originally talking about having a _single_ instance of the accumulator.
I was confused because we need to be able to have multiple snapshots over time.
But you are talking about a single instance of a given _snapshot_. 

Yes :+1: we don't need to clone this snapshot each time. I'll update this.